File: skill_sdk/tests/fixtures/js-project/src/shop/discount/BogoDiscount.js
import { DiscountStrategy } from './DiscountStrategy.js';

/**
 * Buy-one-get-one — cheapest item is free.
 *
 * @implements {DiscountStrategy}
 */
export class BogoDiscount extends DiscountStrategy {
  /**
   * @param {import('../models/Order.js').Order} order
   * @returns {number}
   */
  apply(order) {
    if (order.getItems().length < 2) {
      return order.rawTotal();
    }
    let cheapest = Number.MAX_VALUE;
    for (const item of order.getItems()) {
      if (item.subtotal() < cheapest) {
        cheapest = item.subtotal();
      }
    }
    return order.rawTotal() - cheapest;
  }
}

File: skill_sdk/tests/fixtures/cpp-project/src/shop/BogoDiscount.cpp
#include "BogoDiscount.h"

double BogoDiscount::Apply(const Order& order) {
    if (order.Items.size() < 2) {
        return order.RawTotal();
    }
    double cheapest = order.Items[0].Subtotal();
    for (const auto& item : order.Items) {
        double s = item.Subtotal();
        if (s < cheapest) {
            cheapest = s;
        }
    }
    return order.RawTotal() - cheapest;
}
